src/main/java/com/epam/reportportal/karate/ReportPortalPublisher.java

```diff
@@ -0,0 +1,257 @@
+package com.epam.reportportal.karate;
+
+import com.epam.reportportal.listeners.ListenerParameters;
+import com.epam.reportportal.service.Launch;
+import com.epam.reportportal.service.ReportPortal;
+import com.epam.ta.reportportal.ws.model.FinishExecutionRQ;
+import com.epam.ta.reportportal.ws.model.FinishTestItemRQ;
+import com.epam.ta.reportportal.ws.model.StartTestItemRQ;
+import com.epam.ta.reportportal.ws.model.launch.StartLaunchRQ;
+import com.epam.ta.reportportal.ws.model.log.SaveLogRQ;
+import com.intuit.karate.core.*;
+import io.reactivex.Maybe;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Calendar;
+import java.util.Date;
+import java.util.HashSet;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Supplier;
+
+import static com.google.common.base.Strings.isNullOrEmpty;
+import static org.apache.commons.lang3.StringUtils.isNotBlank;
+
+public class ReportPortalPublisher {
+    private static final Logger LOGGER = LoggerFactory.getLogger(ReportPortalPublisher.class);
+    private static final String INFO_LOG_LEVEL = "INFO";
+    private static final String ERROR_LOG_LEVEL = "ERROR";
+    private static final String WARN_LOG_LEVEL = "WARN";
+    private static final String PASSED = "PASSED";
+    private static final String FAILED = "FAILED";
+    private static final String SKIPPED = "SKIPPED";
+    private static final String STOPPED = "STOPPED";
+    private static final String RESETED = "RESETED";
+    private static final String CANCELLED = "CANCELLED";
+
+    private final Supplier<Launch> launch;
+    private final ConcurrentHashMap<String, Maybe<String>> featureIdMap;
+    private final ConcurrentHashMap<String, Maybe<String>> scenarioIdMap;
+    private final ConcurrentHashMap<Maybe<String>, Long> stepStartTimeMap;
+    private Maybe<String> stepId;
+
+    public ReportPortalPublisher() {
+        this(createLaunchSupplier());
+    }
+
+    public ReportPortalPublisher(Supplier<Launch> launch) {
+        this.launch = launch;
+        this.featureIdMap = new ConcurrentHashMap<>();
+        this.scenarioIdMap = new ConcurrentHashMap<>();
+        this.stepStartTimeMap = new ConcurrentHashMap<>();
+    }
+
+    private static Supplier<Launch> createLaunchSupplier() {
+        return new Supplier<>() {
+            private final Launch launch = createLaunch();
+
+            @Override
+            public Launch get() {
+                return launch;
+            }
```

Comment:
Do you want it to create a Launch on the class initialization or on the first call to `get` method? I'm asking that because in your case it will create a launch even if there is no features to run. Not all users like that. If you would create Launch on the first `get` call it won't create an empty Launch.

Reply:
Refactored. Fixed.